libgadget/cooling_qso_lightup.c

```diff
@@ -0,0 +1,622 @@
+    tw->ngbiter_type_elsize = sizeof(TreeWalkNgbIterBase);
+    tw->ngbiter = ionize_ngbiter;
+
+    tw->fill = (TreeWalkFillQueryFunction) ionize_copy;
+    tw->reduce = NULL;
+    tw->postprocess = NULL;
+    tw->query_type_elsize = sizeof(TreeWalkQueryBase);
+    tw->result_type_elsize = sizeof(TreeWalkResultBase);
+
+    struct QSOPriv priv[1];
+    priv[0].spin = init_spinlocks(PartManager->NumPart);
+    priv[0].fof = fof;
+    /* Ionization counters*/
+    priv[0].N_ionized = ta_malloc("n_ionized", int64_t, omp_get_max_threads());
+    memset(priv[0].N_ionized, 0, sizeof(int64_t) * omp_get_max_threads());
+
+    tw->priv = priv;
+
+    /* This runs only on one BH*/
+    if(qso_ind > 0)
+        treewalk_run(tw, &qso_cand[qso_ind], 1);
+    else
+        treewalk_run(tw, NULL, 0);
+
+    int64_t N_ionized;
+    int i;
+    for(i = 0; i < omp_get_max_threads(); i++)
+        N_ionized += priv[0].N_ionized[i];
+
+    ta_free(priv[0].N_ionized);
+
+    return N_ionized;
+}
+
+/* Sequentially turns on quasars.
+ * Keeps adding new quasars until need_more_quasars() returns 0.
+ */
+static void
+turn_on_quasars(double redshift, FOFGroups * fof, ForceTree * tree)
+{
+    int ncand;
+    int * qso_cand = NULL;
+    int64_t n_gas_tot=0, tot_n_ionized=0, ncand_tot=0;
+    sumup_large_ints(1, &SlotsManager->info[0].size, &n_gas_tot);
+    double atime = 1./(1 + redshift);
+    double desired_ion_frac = gsl_interp_eval(HeIII_intp, He_zz, XHeIII, atime, NULL);
+    /* If the desired ionization fraction is above a threshold (by default 0.95)
+     * ionize all particles*/
+    if(desired_ion_frac > QSOLightupParams.heIIIreion_finish_frac) {
+        int i, nionized=0;
+        int64_t nion_tot=0;
+        #pragma omp parallel for reduction(+: nionized)
+        for (i = 0; i < PartManager->NumPart; i++){
+            if (P[i].Type == 0)
+                nionized += ionize_single_particle(i);
+        }
+        sumup_large_ints(1, &nionized, &nion_tot);
+        message(0, "HeII: Helium ionization finished, flash-ionizing %ld particles (%g of total)\n", nion_tot, (double) nion_tot /(double) n_gas_tot);
+    }
+
+    double rhobar = All.CP.OmegaBaryon * (3 * HUBBLE * All.CP.HubbleParam * HUBBLE * All.CP.HubbleParam)/ (8 * M_PI * GRAVITY) / pow(1 + redshift,3);
+    double totbubblegasmass = 4 * M_PI / 3. * pow(QSOLightupParams.mean_bubble, 3) * rhobar;
+    /* Total expected ionizations if the bubbles do not overlap at all
+     * and the bubble is at mean density.*/
+    int64_t non_overlapping_bubble_number = n_gas_tot * totbubblegasmass / All.CP.OmegaBaryon;
+    double initionfrac = gas_ionization_fraction();
+    double curionfrac = initionfrac;
+
+    if(curionfrac < desired_ion_frac) {
+        ncand = build_qso_candidate_list(&qso_cand, fof);
+        walltime_measure("/HeIII/Find");
+    }
+
+    while (curionfrac < desired_ion_frac){
+        /* Get a new quasar*/
+        int new_qso = choose_QSO_halo(ncand, &ncand_tot, fof, MPI_COMM_WORLD);
```

Comment:
The MPI gather is only used for detecting if the global index falls into the current rank; so technically you could have generated a list of non-repeating global fof indices, and convert them to local with a precomputed offset table.

Reply:
We could have done that, but it is much more complicated. We can try it if this ever turns into a scalability problem

Reply:
In fact, I will do this!